docs/source/developer/tests.rst

```diff
@@ -117,27 +126,170 @@ test is run while `record_flight_data` is in place.
+    if test.recording:
       time.sleep(10)
+
+Writing Terraform Bootstrapped Tests for AWS Resources
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Note: These instructions are only for recording data in AWS. For Azure, GCP, or
+other cloud providers, see the corresponding documentation for information on how
+to record data there.
+
+The `pytest-terraform <https://github.com/cloud-custodian/pytest-terraform>`_ plugin
+is used to streamline the creation and removal of AWS resources using Terraform during
+functional testing.
+
+Terraform modules should be organized in `tests/terraform`. So for a test named
+`test_file_example` the terraform files for that would exist in `tests/terraform/file_example`.
+
+Here's an example terraform file for the upcoming code example. It's placed in
+`tests/terraform/file_example/main.tf`.
+
+  .. code-block:: terraform
+
+    resource "local_file" "bar" {
+       content = "bar!"
+       filename = "${path.module}/bar.txt"
+    }
+
+When invoked, this terraform module will create a file `bar.txt` with the
+contents ``bar!``.
+
+When first creating a test, explicitly set the `replay` parameter to ``False``.
+This will force terraform to run on each invocation of the test and perform
+the flight recording function.
+
+Given the above example terraform file, the following code example is used to
+access data from the terraform decorator, passed in as the same name as the
+first parameter, in this case `file_example`
+
+  .. code-block:: python
+    :emphasize-lines: 1
+
+    @terraform('file_example', replay=False)
+    def test_file_example(file_example):
+
+        assert file_example['local_file.bar.content'] == 'bar!'
+
+Now run this test using the following command to generate terraform flight
+recordings:
+
+  .. code-block:: shell
+
+    $ pytest tests/path/to/test.py -s -v
+
+This may take a little while as tests are typically contacting AWS.
+All Terraform state is recorded in the same directory of the terraform module
+as a `tf_resources.json` file.
+
+  .. code-block:: shell
+
+    $ ls tests/terraform/file_example/
+    main.tf
+    tf_resources.json
+
+Each invocation of the test where ``replay=False`` is set the `tf_resources.json`
+contents are replaced and updated with that runs output.
+
+When the test is completed, remove ``replay=False``:
+
+  .. code-block:: python
+    :emphasize-lines: 1
+
+    @terraform('file_example')
+    def test_file_example(file_example):
+
+        assert file_example['local_file.bar.content'] == 'bar!'
+
+Now when the test is run it will use the data previously recorded terraform
+resources and not run terraform directly. When committing your test, don't
+forget to include the `tests/terraform/file_example` directory!
+
+Controlling Resource Cleanup
+============================
+
+By default, pytest-terraform will run ``terraform destroy`` after each wrapped
+test function. For majority of use cases this will likely be okay. If terraform
```

Comment:
explain why it wouldn't be.. ie. the policy test execution would delete provisioned resources from terraform.